File: src/2022-intern/799.champagne-tower.js
/*
 * @lc app=leetcode id=799 lang=javascript
 *
 * [799] Champagne Tower
 */

// @lc code=start
/**
 * @param {number} poured
 * @param {number} query_row
 * @param {number} query_glass
 * @return {number}
 */
var champagneTower = function (poured, query_row, query_glass) {
  const layers = new Array(query_row + 2);
  for (let i = 0; i < layers.length; i++) {
    layers[i] = new Array(i + 1);
    layers[i].fill(0);
  }

  layers[0][0] = poured;

  for (let i = 0; i <= query_row; i++) {
    for (let j = 0; j <= i; j++) {
      if (layers[i][j] > 1) {
        const overflow = layers[i][j] - 1;
        layers[i][j] = 1;
        layers[i + 1][j] += overflow / 2;
        layers[i + 1][j + 1] += overflow / 2;
      }
    }
  }

  return layers[query_row][query_glass];
};
// @lc code=end

champagneTower(2, 1, 1);

File: src/2026/799.champagne-tower.cpp
/*
 * @lc app=leetcode id=799 lang=cpp
 *
 * [799] Champagne Tower
 */

#include <algorithm>
#include <vector>

// @lc code=start
class Solution {
 public:
  double champagneTower(int poured, int query_row, int query_glass) {
    std::vector<std::vector<double>> tower(2);
    tower[0].resize(1);
    tower[0][0] = poured;
    for (int row = 0; row < query_row; row++) {
      // Pour into next level cups
      std::ranges::fill(tower[(row + 1) % 2], 0.0);
      tower[(row + 1) % 2].resize(row + 2);
      for (int col = 0; col < tower[row % 2].size(); col++) {
        const double excess_liquid = std::max(0.0, tower[row % 2][col] - 1.0);
        tower[row % 2][col] -= excess_liquid;
        tower[(row + 1) % 2][col] += excess_liquid / 2;
        tower[(row + 1) % 2][col + 1] += excess_liquid / 2;
      }
    }

    return std::min(1.0, tower[query_row % 2][query_glass]);
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.champagneTower(100000009, 33, 17);
}